PR: https://github.com/odoo/odoo/pull/211043

From: f2965048f60fe6c815b3e50fa714c97a93dfb5d3
From: Yagnik ✗ (yagp)
Date: 2025-12-19 12:08:14

Breaking data model changes scores: del:1.4 + add:0, change matches:
-    is_company = fields.Boolean(string='Is a Company', default=False,
+    is_company = fields.Boolean(string='Is a Company', default=False, compute="_compute_is_company", store=True,
-    company_type = fields.Selection(string='Company Type',

Total Changes: 89

[REM] base: company_type is dead for good

This commit removes the 'company_type' field, which was previously used to
distinguish between 'Person' and 'Company' contacts.

The goal is to simplify the contact form and model by eliminating this
distinction, as part of an effort to streamline and simplify the
contact information structure.

This change is in line with the broader initiative to remove legacy
fields and make the model more straightforward as possible.

is_company becomes a computed stored field, allowing each l10n to
specify how a company is defined in a sepcific country. By default
we will now rely on VAT and commercial partner to compute `is_company`.

Removed fields:
base  : company_type, company_name
crm : commercial_partner_id

Task-4714892

Part-of: odoo/odoo#211043
Related: odoo/enterprise#86089
Related: odoo/upgrade#7724
Related: odoo/upgrade-util#290
Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/odoo/addons/base/models/res_company.py
+++ b/odoo/addons/base/models/res_company.py
@@ -294,7 +294,6 @@ class ResCompany(models.CachedModel):
             partners = self.env['res.partner'].with_context(default_parent_id=False).create([
                 {
                     'name': vals['name'],
-                    'is_company': True,
                     'image_1920': vals.get('logo'),
                     'email': vals.get('email'),
                     'phone': vals.get('phone'),

--- a/odoo/addons/base/models/res_partner.py
+++ b/odoo/addons/base/models/res_partner.py
@@ -266,14 +266,10 @@ class ResPartner(models.Model):
         'Formatted Email', compute='_compute_email_formatted',
         help='Format email address "Name <email@domain>"')
     phone = fields.Char()
-    is_company = fields.Boolean(string='Is a Company', default=False,
+    is_company = fields.Boolean(string='Is a Company', default=False, compute="_compute_is_company", store=True,
         help="Check if the contact is a company, otherwise it is a person")
     is_public = fields.Boolean(compute='_compute_is_public', compute_sudo=True)
     industry_id: ResPartnerIndustry = fields.Many2one('res.partner.industry', 'Industry')
-    # company_type is only an interface field, do not use it in business logic
-    company_type = fields.Selection(string='Company Type',
-        selection=[('person', 'Person'), ('company', 'Company')],
-        compute='_compute_company_type', inverse='_write_company_type')
     company_id: ResCompany = fields.Many2one('res.company', 'Company', index=True)
     color = fields.Integer(string='Color Index', default=0)
     user_ids: ResUsers = fields.One2many('res.users', 'partner_id', string='Users', bypass_search_access=True)
@@ -374,14 +370,14 @@ class ResPartner(models.Model):
         type_description = dict(self._fields['type']._description_selection(self.env))
 
         name = self.name or ''
-        if self.company_name or self.parent_id:
+        if self.parent_id:
             if not name and self.type in displayed_types:
                 name = type_description[self.type]
             if not self.is_company and not self.env.context.get('partner_display_name_hide_company'):
                 name = f"{self.commercial_company_name or self.sudo().parent_id.name}, {name}"
         return name.strip()
 
-    @api.depends('is_company', 'name', 'parent_id.name', 'type', 'company_name', 'commercial_company_name')
+    @api.depends('is_company', 'name', 'parent_id.name', 'type', 'commercial_company_name')
     def _compute_complete_name(self):
         for partner in self:
             partner.complete_name = partner.with_context({})._get_complete_name()
@@ -407,7 +403,7 @@ class ResPartner(models.Model):
     @api.depends('parent_id')
     def _compute_user_id(self):
         """ Synchronize sales rep with parent if partner is a person """
-        for partner in self.filtered(lambda partner: not partner.user_id and partner.company_type == 'person' and partner.parent_id.user_id):
+        for partner in self.filtered(lambda partner: not partner.user_id and not partner.is_company and partner.parent_id.user_id):
             partner.user_id = partner.parent_id.user_id
 
     @api.depends_context("uid")
@@ -443,7 +439,7 @@ class ResPartner(models.Model):
             # so that you can reactivate it instead of creating a new one, which would lose its history.
             Partner = self.with_context(active_test=False).sudo()
             vats = [partner.vat]
-            should_check_vat = partner.vat and len(partner.vat) != 1
+            should_check_vat = not self._is_vat_void(partner.vat)
 
             if should_check_vat and partner.country_id and 'EU_PREFIX' in partner.country_id.country_group_codes:
                 if partner.vat[:2].isalpha():
@@ -484,8 +480,6 @@ class ResPartner(models.Model):
                 partner.type_address_label = _('Invoice Address')
             elif partner.type == 'delivery':
                 partner.type_address_label = _('Delivery Address')
-            elif partner.type == 'contact' and partner.parent_id:
-                partner.type_address_label = _('Company Address')
             else:
                 partner.type_address_label = _('Address')
 
@@ -498,10 +492,10 @@ class ResPartner(models.Model):
         for partner in self:
             partner.self = partner.id
 
-    @api.depends('is_company', 'parent_id.commercial_partner_id')
+    @api.depends('parent_id.commercial_partner_id', 'parent_id')
     def _compute_commercial_partner(self):
         for partner in self:
-            if partner.is_company or not partner.parent_id:
+            if not partner.parent_id:
                 partner.commercial_partner_id = partner
             else:
                 partner.commercial_partner_id = partner.parent_id.commercial_partner_id
@@ -562,12 +556,12 @@ class ResPartner(models.Model):
             return
         result = {}
         partner = self._origin
-        if (partner.type or self.type) == 'contact':
+        parent_address = self.parent_id._get_address_values()
+        if (partner.type or self.type) == 'contact' and bool(parent_address):
             # for contacts: copy the parent address, if set (aka, at least one
             # value is set in the address: otherwise, keep the one from the
             # contact)
-            if address_values := self.parent_id._get_address_values():
-                result['value'] = address_values
+            result['value'] = parent_address
         return result
 
     @api.onchange('country_id')
@@ -618,19 +612,6 @@ class ResPartner(models.Model):
                     partner.email
                 ))
 
-    @api.depends('is_company')
-    def _compute_company_type(self):
-        for partner in self:
-            partner.company_type = 'company' if partner.is_company else 'person'
-
-    def _write_company_type(self):
-        for partner in self:
-            partner.is_company = partner.company_type == 'company'
-
-    @api.onchange('company_type')
-    def onchange_company_type(self):
-        self.is_company = (self.company_type == 'company')
-
     @api.constrains('barcode')
     def _check_barcode_unicity(self):
         for partner in self:
@@ -758,6 +739,11 @@ class ResPartner(models.Model):
         if values.get('parent_id') or values.get('type') == 'contact':
             # 1a. Commercial fields: sync if parent changed
             if values.get('parent_id'):
+                address = self._get_address_values()
+                parent_address = self.parent_id._get_address_values()
+                if self.type == 'contact' and bool(address) and bool(parent_address) and address != parent_address:
+                    # If current partner has a different address set, change it's type so we don't lost it
+                    self.type = 'other'
                 self.sudo()._commercial_sync_from_company()
             # 1b. Address fields: sync if parent or use_parent changed *and* both are now set
             if self.parent_id and self.type == 'contact':
@@ -826,6 +812,18 @@ class ResPartner(models.Model):
             website = url.replace(scheme='http').to_url()
         return website
 
+    @api.depends('vat', 'commercial_partner_id')
+    def _compute_is_company(self):
+        """ By default, a partner is considered as a company if they are their own
+        commercial entity (see computed field), and if their VAT is considered as being
+        valid, based on a default heuristic (not void, '/', 'na' or 'NA').
+
+        Each localization can then further refine this definition according to legal
+        definition of what is a company (e.g. more strict VAT, specific field usage,
+        ...) """
+        for partner in self:
+            partner.is_company = partner.commercial_partner_id == partner and not partner._is_vat_void(partner.vat)
+
     def _compute_is_public(self):
         for partner in self.with_context(active_test=False):
             users = partner.user_ids
@@ -856,8 +854,6 @@ class ResPartner(models.Model):
                                             'Linked active users :\n%(names)s', names=", ".join([u.display_name for u in users])))
         if vals.get('website'):
             vals['website'] = self._clean_website(vals['website'])
-        if vals.get('parent_id'):
-            vals['company_name'] = False
 
         # filter to keep only really updated values -> field synchronize goes through
         # partner tree and we should avoid infinite loops in case same value is
@@ -882,12 +878,7 @@ class ResPartner(models.Model):
                             self.env._("The selected company is not compatible with the companies of the related user(s)"))
                 if partner.child_ids:
                     partner.child_ids.write({'company_id': company_id})
-        result = True
-        # To write in SUPERUSER on field is_company and avoid access rights problems.
-        if 'is_company' in vals and not self.env.su and self.env.user.has_group('base.group_partner_manager'):
-            result = super(ResPartner, self.sudo()).write({'is_company': vals.get('is_company')})
-            del vals['is_company']
-        result = result and super().write(vals)
+        result = super().write(vals)
         for partner, pre_values in zip(self, pre_values_list, strict=True):
             if internal_users := partner.user_ids.filtered(lambda u: u._is_internal() and u != self.env.user):
                 internal_users.check_access('write')
@@ -903,8 +894,6 @@ class ResPartner(models.Model):
         for vals in vals_list:
             if vals.get('website'):
                 vals['website'] = self._clean_website(vals['website'])
-            if vals.get('parent_id'):
-                vals['company_name'] = False
         partners = super().create(vals_list)
         # due to ir.default, compute is not called as there is a default value
         # hence calling the compute manually
@@ -919,6 +908,11 @@ class ResPartner(models.Model):
             partner._fields_sync(vals)
         return partners
 
+    def _is_vat_void(self, vat):
+        if not vat:
+            return True
+        return vat in ['/', 'na', 'NA']
+
     @api.ondelete(at_uninstall=False)
     def _unlink_except_user(self):
         users = self.env['res.users'].sudo().search([('partner_id', 'in', self.ids)])
@@ -1002,7 +996,7 @@ class ResPartner(models.Model):
                 'target': 'current',
                 }
 
-    @api.depends('complete_name', 'email', 'vat', 'state_id', 'country_id', 'commercial_company_name')
+    @api.depends('complete_name', 'email', 'vat', 'state_id', 'country_id')
     @api.depends_context(
         'show_address', 'partner_show_db_id',
         'show_email', 'show_vat', 'lang', 'formatted_display_name'
@@ -1012,8 +1006,8 @@ class ResPartner(models.Model):
         for partner in self:
             if partner.env.context.get("formatted_display_name"):
                 name = partner.name or ''
-                if partner.parent_id or partner.company_name:
-                    name = (f"{partner.company_name or partner.parent_id.name} \t "
+                if partner.parent_id:
+                    name = (f"{partner.parent_id.name} \t "
                             f"--{partner.name or type_description.get(partner.type, '')}--")
 
                 if partner.env.context.get('show_email') and partner.email:
@@ -1154,14 +1148,14 @@ class ResPartner(models.Model):
             'state_name': self.state_id.name or '',
             'country_code': self.country_id.code or '',
             'country_name': self._get_country_name(),
-            'company_name': self.commercial_company_name or '',
+            'parent_name': self.commercial_company_name or '',
         })
         for field in self._formatting_address_fields():
             args[field] = self[field] or ''
         if without_company:
-            args['company_name'] = ''
-        elif self.commercial_company_name:
-            address_format = '%(company_name)s\n' + address_format
+            args['parent_name'] = ''
+        elif self.parent_id:
+            address_format = '%(parent_name)s\n' + address_format
         return address_format, args
 
     def _display_address(self, without_company=False):
@@ -1180,7 +1174,7 @@ class ResPartner(models.Model):
     def _display_address_depends(self):
         # field dependencies of method _display_address()
         return self._formatting_address_fields() + [
-            'country_id', 'company_name', 'state_id',
+            'country_id', 'parent_id', 'state_id',
         ]
 
     @api.model
